### PR TITLE
server: on new connection, handshake and dispatch to different type

### PR DESCRIPTION
**Pull request by tiancaiamao**
there is trick that the first Message is used as handshake packet
that's not good but compatible, we can refact it later.

we'll have connections for snapshot, which begin with MessageTypeNone
followed by raw data

### CODE REVIEW

File: src/server/conn.rs [hunk 1]
```diff
@@ -114,45 +123,102 @@ impl Conn {
...
+              S: StoreAddrResolver
+    {
+        let msg_or_none = try!(self.read_one_message());
+        if msg_or_none.is_none() {
+            return Ok(vec![]);
+        }
+
+        let msg = msg_or_none.unwrap();
+        match msg.get_msg_type() {
+            MessageType::None => {
```
Comment (siddontang) [inline src/server/conn.rs]: You must define a new type for snapshot, not None.

Comment (tiancaiamao) [inline src/server/conn.rs]: Yes, but we can do it later. 
That's a minimal change and don't involve kvproto.

Comment (siddontang) [inline src/server/conn.rs]: Adding a new type in kvproto doesn't cause any problem. 
None is very bad, what is None for? An invalid message or Snapshot?

File: src/server/conn.rs [hunk 2]
```diff
@@ -114,45 +123,102 @@ impl Conn {
...
+            }
+        }
+    }
+
+    fn read_snapshot<T, S>(&mut self, _: &mut EventLoop<Server<T, S>>) -> Result<Vec<ConnData>>
+        where T: RaftStoreRouter,
+              S: StoreAddrResolver
+    {
+        // TODO
+        Ok(vec![])
```
Comment (siddontang) [inline src/server/conn.rs]: `unimplemented!()`

File: src/server/mod.rs [hunk 3]
```diff
@@ -136,7 +136,9 @@ impl Display for ConnData {
             MessageType::CopResp => write!(f, "[{}] coprocessor response", self.msg_id),
             MessageType::PdReq => write!(f, "[{}] pd request", self.msg_id),
             MessageType::PdResp => write!(f, "[{}] pd response", self.msg_id),
-            MessageType::None => write!(f, "[{}] invalid message", self.msg_id),
+            MessageType::None | MessageType::Snapshot => {
```
Comment (siddontang) [inline src/server/mod.rs]: print Snapshot here.

File: src/server/conn.rs (hunk 2 above)
Comment (tiancaiamao) [inline src/server/conn.rs]: actually we never run here in this PR, because client haven't changed yet

Comment (siddontang) [inline src/server/conn.rs]: so use unimplemented.

File: src/server/mod.rs [hunk 4]
```diff
@@ -136,6 +136,7 @@ impl Display for ConnData {
             MessageType::CopResp => write!(f, "[{}] coprocessor response", self.msg_id),
             MessageType::PdReq => write!(f, "[{}] pd request", self.msg_id),
             MessageType::PdResp => write!(f, "[{}] pd response", self.msg_id),
+            MessageType::Snapshot => write!(f, "[{}] pd snapshot", self.msg_id),
```
Comment (siddontang) [inline src/server/mod.rs]: remove pd

**Comment by tiancaiamao** [comment]
will be include in https://github.com/pingcap/tikv/pull/648

**Pull request closed without merging**